packages/liveblocks-react-ui/src/components/Composer.tsx

```diff
@@ -555,9 +570,53 @@ export const Composer = forwardRef(
     forwardedRef: ForwardedRef<HTMLFormElement>
   ) => {
     const client = useClient();
-    const createThread = useCreateThread();
-    const createComment = useCreateComment();
-    const editComment = useEditComment();
+    const thread = useContext(ThreadDataContext);
+    const comment = useContext(CommentDataContext);
+    const room = useRoomOrNull();
+
+    function getRoomId(): string {
+      // Both threadId and commentId are provided (editing a comment)
+      if (threadId !== undefined && commentId !== undefined) {
+        if (comment && comment.id === commentId) {
+          return comment.roomId;
+        } else if (thread && thread.id === threadId) {
+          return thread.roomId;
+        } else if (room) {
+          return room.id;
+        } else {
+          throw new Error(
+            "Composer is provided both threadId and commentId but is not a descendant of Comment, Thread, or RoomProvider component"
+          );
+        }
+      }
+      // Only threadId is provided (replying to a thread)
+      else if (threadId !== undefined) {
+        if (thread && thread.id === threadId) {
+          return thread.roomId;
+        } else if (room) {
+          return room.id;
+        } else {
+          throw new Error(
+            "Composer is provided a threadId but is not a descendant of Thread or RoomProvider component"
+          );
+        }
+      }
+      // Default to room context (creating a new thread)
+      else if (room) {
+        return room.id;
+      }
+
+      // No valid context found
+      throw new Error(
+        "Composer must be a descendant of RoomProvider component"
+      );
+    }
```

Comment:
This `getRoomId()` method (in `Composer` component) does a few things to figure out which room id to use. 

In summary, if `commentId` and `threadId` are both passed to the component, it tries to obtain the room id from the comment context. If comment data is not found, then it tries to obtain the room id from the thread context. If neither are found, it tries to look at the room context.

1. In the following example, `Composer` is a descendant of `Comment`, so the composer will use the room id obtained from the comment context.

    ```tsx
    <Comment comment={comment}>
      <Composer commentId={comment.id} threadId={thread.id} /> 
    </Comment />
    ```

2. In the following example, `Composer` is a descendant of `Thread` but not of `Comment`, so the composer will use the room id obtained from the thread context.

    ```tsx
    <Thread thread={thread}>
      <Composer threadId={thread.id} /> 
    </Comment />
    ```

3. In the following example, `Composer` neither a descendant of `Thread` nor of `Comment`, so the composer will use the room id obtained from the RoomProvider context.

    ```tsx
    <RoomProvider roomId={roomId}>
      <Composer /> 
    </Comment />
    ```

If none of `RoomProvider`, `Thread` or `Comment` context were found, we throw an error.

We also do some extra check to ensure that even if a `Composer` is a descendant of a `Thread` component, the thread id of the thread context must match the `threadId` passed to the `Composer`, otherwise, we fall back to the room context.

Reply:
Wow, a pretty original solution that I did not expect!

I can see how this solution works indeed for the problem at hand, and how it helped you land on the implementation. But I do worry a bit about relying on React contexts for too many things. My areas of worry are: authority, consistency, performance, and (future) use outside of React.

**Authority**
Currently, we rely on a parent component to set/provide the right value. If that component somehow sets an invalid value, or ever fails to update or reset the value correctly in certain edge cases.

**Consistency**
It's technically possible to create a state like this (bare with me if I didn't get it exactly right here):

```tsx
<ComposerRoomIdContext value="room-a">
  <ThreadDataContext value="unrelated-thread-id">
  </ThreadDataContext>
</ComposerRoomIdContext>
```

If the `room-a` and `unrelated-thread-id` form (or can form) an invalid "pair", then the category of bugs that would end up as a result would be extremely hard to reason about.

**Performance**
Not the biggest worry, but stacking so many extra components in the component hierarchy is probably not going to help with performance (and ease of understanding).

### How else?

You're way deeper into the weeds here, but from my (maybe naive) outside understanding of this problem, I'd be inclined to let each component/hook simply ask the central authority which `roomId` a certain `threadId` or `commentId` belongs to. This data is readily available from the pool maintained by the Umbrella store already. Would this be an option here? I feel like it could make this implementation much simpler.

I'm also on this path of thought because of this somewhat loosely related issue LB-1379 we also need to solve for.

So what about we would allow these hooks to do something like this instead?

```tsx
const client = useClient();
client[kInternal].getRoomIdForResource("th_123");  // or
client[kInternal].getRoomIdForResource("cm_123");
```

(Or we make those two lookups, I don't care.) But the point is to ask the authority basically. These lookups could be indexed, and would be super fast O(1), so without any performance overhead or all that component tree stacking. And it would always be consistent/correct, no matter how your state changes.

Happy to talk more about this!